### DISCUSSION TITLE
[Feature] Client-side checksum computation

### DISCUSSION THREAD

**Discussion by ben-z** [immich-app/immich#4154] [Feature Request]
### The feature

Currently, it appears that the SHA1 checksum is calculated and deduplication happens on the server side, after the upload ([1](https://github.com/immich-app/immich/pull/525), [2](https://github.com/immich-app/immich/pull/557)). This can take a long time (and takes a lot of bandwidth) if there are multiple devices that share the library (e.g. iCloud Photo Library). Is there a way to implement client-side SHA1 computation, and check for duplicates before the file is uploaded? I feel that this would significantly speed up backing up the same library from multiple devices.

There's some discussion on this in the comments of https://github.com/immich-app/immich/discussions/1323

### Platform

- [ ] Server
- [X] Web
- [X] Mobile

**Comment by jrasm91** [comment]
Sha1 on the web is quite a pain and pretty inefficient and/or complex. There is an option to prehash in the CLI before upload, but the server will still compute it if it ends up being uploaded. And actually I think this only exists in the unreleased CLI in the main repo. In the mobile app, hash to determine if it should be uploaded or not is currently a work in progress.

**Reply by ben-z** [reply]
Thanks for the info! For the record, I just came across the CLI client side hashing pR: https://github.com/immich-app/CLI/pull/88

**Reply by etnoy** [reply]
the new v2 cli supports client side hashing

**Reply by ben-z** [reply]
Awesome! For the record, here are the tickets to track mobile client-side hashing:
- https://github.com/immich-app/immich/issues/2567
- https://github.com/immich-app/immich/issues/1553
